### PR TITLE
Feat: git repos integration

### PR DESCRIPTION
**Pull request by chrismclarke**
PR Checklist

- [x] - Latest `master` branch merged
- [x] - PR title descriptive (can be used in release notes)

## Description
Rework of #1408 to try resolve significant merge conflicts and reduce the total diffs for easier testing/review

**Deployment scripts**
```
yarn workflow deployment import
```
Import a deployment from an external git repo

**Deployment repos**
Deployment data now can be stored in an external app repo and managed locally. A minimal git management system has been added to handle the process of pushing and pulling data from the remote repo, as well as creating release branches

For any repo configured with an external git repo, both the `deployment set` and `deployment sync` commands will fetch latest data from the remote repo first

```
yarn workflow repo publish
```
Tag the active app content and publish to github. Follows semantic versioning and creates a new branch on the content repo for review and merge

**Additional Changes**
A number of scripts have been added to enable the new git and deployment workflows. Many other scripts have been modified to function within changed deployment contexts (e.g. processes to import data from within the app), and various bits of code tidying applied where identified.

Additionally a number of work-in-progress github actions have been developed for use in standalone content repos to build/deploy the full app from a deployment repo. These can be seen at:
https://github.com/c-squared-dev/idems-app-debug-git/tree/main/.github/workflows

## Review Notes
Requires `yarn install`

**Main Test Cases**

0. Creating a new repo for testing
It would be useful to create a new git repo managed by IDEMS to store a debug deployment, alongside google drive sheet and assets folders. Ideally this repo could be the basis for the future storing of debug sheets or similar.
I've created personal repos for demo purposes (used below)

2. Importing external repo
```
yarn workflow deployment import https://github.com/c-squared-dev/idems-app-debug-git
```
This should import and populate a testing repo created

3. Update content and sync
```
yarn workflow sync
```
(note, if using personal repo may be issue with gdrive permissions)

4. Publishing external repo
```
yarn workflow repo publish
```
This should indicate changes from sync action above and prompt for version update
(note, if using personal repo may be issue with github repo permissions)

**Additional Test Cases**
- Review documentation updates
- See example github actions added for personal deployment repo (will want to migrate more actions in the future)

## Future TODOs 
**Core Functionality**
- [ ] Fix CI integrations

**Code Tidying**
- [ ] Add outputs for deployment create/import and use for set

**CI**
- [ ] Add example actions for standalone deployment (web preview, android release)
- [ ] Add scripts and templates to make it easier to automatically populate deployment repo actions
- [ ] Add actions to automatically open PRs

**Content**
- [ ] PR to remove app-data (once fully deprecated), should also change default app_data.output_path
- [ ] Populate PLH content repo
- [ ] Remove hardcode app_data paths from config

**Repo**
- [ ] Remove historic large file history and lfs files
- [ ] Handle line endings (gitattributes?)
- [ ] Ensure git-lfs installed

**Scripts**
- [ ] Possibly refactor script as a provider (e.g. custom provider)
- [ ] Move all workflows to scripts (?)… unlikely to include custom
- [ ] Rename tasks to providers
- [ ] Publish external app scripts package and include in repo with package.json to allow triggering from content repo
- [ ] Migrate additional code to ts-morph (e.g. template file)

**Git**
- [ ] Scripts to populate a specific content version/track stable/preview etc. pipelines or branches
- [ ] Add support for target app version from content repo and include in build

**Misc**
- [ ] Update release system for core app
- [ ] Consider an extension system to manage additional deployment settings, e.g. android folder containing splash/icon for generating android assets or capacitor settings (populated on deployment change)
- [ ] Better feedback on long operations, such as git repo sync for large repo

## Git Issues

Closes #1408

## Screenshots/Videos
**Publish content release via `yarn workflow repo publish`**

image

**Updated deployment docs (see Github Management)**
127 0 0 1_8000_developers_deployments_ (2)

### CODE REVIEW

**Comment by jfmcquade** [comment]
This is great, @chrismclarke, it's very pleasing seeing the different systems interact. I've been running some functional testing and have a few notes below.

Firstly, and I think unrelated tot he content of the PR, the unit tests are failing when running a test build, see [this action](https://github.com/IDEMSInternational/parenting-app-ui/actions/runs/4991120439/jobs/8937193558?pr=1755).
  - This appears to be since I merged master (which involved resolving a merge conflict in `package.json`)
  - Looking at the differences [here](https://github.com/IDEMSInternational/parenting-app-ui/pull/1755/files/a992c407a30fb0cfb55575ab78991c14d723ada9..26d7d8131c6087e733d76d3ea891782ede326be6), the only changes are to package.json and yarn.lock, so I'm not sure why that's causing the tests to fail

#### Functional testing notes
- I noticed that the asset sizes logged when running `yarn workflow sync` were incorrect. However I think this is an issue on master, I've opened PR #1920 as a fix.

- Pushing to your personal repo was not possible due to github permissions (as you anticipated in your testing notes). A minor improvement would be to improve the error logging in that situation (namely a 403 from github), currently the error appears like this:
    github 403

- I've set up a new repo for a debug deployment, as you suggested, here: [github.com/jfmcquade/app-debug](https://github.com/jfmcquade/app-debug). It can be transferred to the IDEMS org on github, but I don't have permissions to do so. It currently uses you gdrive folders for debugging, but this can be updated later too.

- Updating to reflect changes made to remote deployment
  - This is achieved by re-setting the deployment, e.g. by running `yarn workflow set` and selecting the current deployment.
  - It would seem intuitive to me to have an additional command such as `yarn workflow deployment update` (or perhaps `pull`, or `sync`) that updates the state of the local deployment to reflect the remote deployment repo

- New docs
  - The docs refer to the command `yarn workflow content_release`, which I believe you have replaced with `yarn workflow repo publish`

- For discussion
  - It could be challenging for authors to learn the new workflow until they have a good mental model of the structures.
  - E.g. `yarn workflow sync` vs `yarn worklfow deployment set` to update local content from remote. One pulls from google drive, the other pulls from remote content repo – I understand the difference, but it will need some explanation for newcomers.

**Comment by chrismclarke** [comment]
Thanks for testing @jfmcquade 
Really strange the tests started failing after you merged master into this branch - I can recreate locally but still don't fully understand (seems data pipe operators not parsing args correctly, although that code was unchanged here so seems to be a typescript issue potentially) - will see if recreating the yarn lock fixes

**Comment by chrismclarke** [comment]
@jfmcquade 
Looks like reverting the yarn lock fixes (I did `git restore --source master yarn.lock` and then ran a fresh yarn install)

I've also added a PR to improve the error warning and also automatically rollback a failed publish action (using your repo as a demo for myself)
image

As for the testing repo, I'll leave it to @esmeetewinkel if they want to create/move a testing repo either before or after this PR is merged in

**Comment by chrismclarke** [comment]
@jfmcquade 
I think all this leaves is documentation and clarifications. I'll tidy up the docs a little now, although a couple general points to be aware of:

1. The `yarn workflow sync` action does refresh content from remote repo each time, so I don't think any extra pull/sync required (and is better than setting the repo again). I'll make that clearer in the docs

2. Given that we are now handling repos inside of repos I think mental models are indeed going to be a challenge. really we will need to improve documentation, and longer-term likely create some sort of graphical interface to the scripts themselves to handle tasks like publishing content

**Review by jfmcquade — APPROVED** [review]
> Looks like reverting the yarn lock fixes (I did `git restore --source master yarn.lock` and then ran a fresh yarn install)

Great, the only cause of the issue I can think of at the moment is the fact that I manually resolved the merge conflicts in `package.json` and didn't alphabetise the list of dependencies. Then I ran `yarn install` and committed `yarn.lock` and the newly alphabetised `package.json` at different points, so the two files in the repo were in some sense out of step with each other. If this caused a problem then it was a very strange manifestation, but I'm not going to worry about it and good to have that command for future reference.


> I've also added a PR to improve the error warning and also automatically rollback a failed publish action

Looks good, it makes sense to keep the output from git but I think it's nice to have it handled and styled to mitigate against authors thinking that "it's broken".


> The `yarn workflow sync` action does refresh content from remote repo each time, so I don't think any extra pull/sync required (and is better than setting the repo again). I'll make that clearer in the docs

Ah yes, I didn't spot this. That makes sense, and means that the complications for authors of choosing when to sync from google drive and when to sync from the content repo are less of a concern. 

I'd be happy to contribute to to improving documentation, I think it'd be good to work with the new system for a bit and see what arises from that.

**Review by esmeetewinkel — APPROVED** [review]
Functional test passed. As requested in the notes I've created a debug content repo https://github.com/IDEMSInternational/App-Debug-Content and moved the quality assurance sheets and assets there. I've been able to run the app with this content. 

[Edited] I'm still figuring out how to effectively work on content with the new system. Before this PR, I would
1. create a content branch
2. a. work on some spreadsheets 
    b. sync content and run app
    c. look at the changes in `.json` files and in app
3. commit into content branch 
4. create PR from content branch into master
5. (review and) merge

There are **many** iterations of 2a+2b+2c and at least a couple of 2+3 before I would finally be happy with the changes and create a PR (which was having no review up to now, but should ideally have review).

**Comment by esmeetewinkel** [comment]
non-blocking: for me the pull request link generated when running `yarn workflow repo publish` does not work when the link wraps onto multiple line because the terminal window is narrow 
image

**Comment by esmeetewinkel** [comment]
Merging with some outstanding (nit) comments, so that I can move forward reviewing other PRs that require testing on the now separated debug deployment

File: documentation/docs/developers/deployments.md
Comment (chrismclarke) [reply documentation/docs/developers/deployments.md:104]: Included in #1959

**Comment by chrismclarke** [comment]
@esmeetewinkel 

> At first I couldn't see the changes to the content .json files anywhere, but (I think) yarn workflow deployment open did the trick. How do I go back to my "actual" parenting-app-ui repo and it's branches? yarn workflow deployment close would make sense to me. N.B. open does not appear in the list of available commands.

VScode can be a bit confusing because essentially we are opening a repo within a repo. The `open` command simply opens a new instance of VSCode inside the inner repo (deployment folder). So you should then have two separate vscode windows, and can just close the latter one when no longer required.

There is also some additional funcationality in VSCode to show both git repos in the main source control panel, which usually appears if using a deployment that is hidden from main repo (e.g. the idems app debug repo)
image

You could work directly within this double source control, or simply ignore by right-clicking and saying `Close Repository`

> non-blocking: for me the pull request link generated when running yarn workflow repo publish does not work when the link wraps onto multiple line because the terminal window is narrow

Resolved with #1959

**Pull request merged**
